### PR TITLE
[FIX] `website_sale`: prevent paying a cancelled website order

### PR DESCRIPTION
**Pull request by ivantodorovich**
**Steps to reproduce:**

1. Add a few products to your cart.
2. Go to the checkout page, and stop right before clicking on 'Pay now'
   (ie: customer goes to find the credit card and comes back later)
3. On another tab, go to the backend and cancel the sale.order
4. Go back to the website tab, and click on Pay now
5. Complete the payment process

**Result:**

* The customer is able to pay a cancelled order.

**Expected:**

* It shouldn't be possible to click on 'Pay now'

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ivantodorovich** [comment]
ping @AntoineVDV

**Review by petrus-v — APPROVED** [review]

**Review by AntoineVDV — COMMENTED** [review]
I don't think that this is the right place for addressing the issue, even if it is indeed "fixed" with the validation error.
In the `payment` module, we [check the state of the invoice to prevent customers from paying it if it's canceled](https://github.com/odoo/odoo/blob/a26ff947ff2b499d2371ea8a777627824adc26db/addons/payment/controllers/portal.py#L111-L113) before entering the payment flow and crashing it mid-course.
In `website_sale` I see that the `_get_shop_payment_values` method can include an `errors` item in the rendering values to prevent the payment entirely, would that be a suitable option for you @Williambraecky?

**Comment by ivantodorovich** [comment]
The thing is `_get_shop_payment_values` is called when rendering the `shop/payment` page.
The user can still leave to get his wallet when on this page, AFTER it's rendered (and thus after this method is called).

FTR, raising the ValidationError here doesn't look bad:
image

**Comment by ivantodorovich** [comment]
In fact, the `/shop/payment` controller will redirect to `/shop` if the cart is cancelled, so it's not needed there. (just tested it)

- https://github.com/odoo/odoo/blob/a26ff947ff2b499d2371ea8a777627824adc26db/addons/website_sale/controllers/main.py#L1062

**Comment by Williambraecky** [comment]
> I don't think that this is the right place for addressing the issue, even if it is indeed "fixed" with the validation error. In the `payment` module, we [check the state of the invoice to prevent customers from paying it if it's canceled](https://github.com/odoo/odoo/blob/a26ff947ff2b499d2371ea8a777627824adc26db/addons/payment/controllers/portal.py#L111-L113) before entering the payment flow and crashing it mid-course. In `website_sale` I see that the `_get_shop_payment_values` method can include an `errors` item in the rendering values to prevent the payment entirely, would that be a suitable option for you @Williambraecky?

Honestly according to what @ivantodorovich is saying, the fix seems ok to me.
I also assume this error is quite rare.

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -1249,6 +1249,10 @@ def shop_payment_transaction(self, order_id, access_token, **kwargs):
         except AccessError:
             raise ValidationError("The access token is invalid.")
 
+        order = request.env['sale.order'].browse(order_id)
```
Comment (AntoineVDV) [inline addons/website_sale/controllers/main.py:1252]: Test for the validity of the id first, so it crashes ASAP if it needs to. This way we're certain to read the actual state of the order.
```suggestion
        order = request.env['sale.order'].browse(order_id).exists()
```

File: addons/website_sale/controllers/main.py [hunk 2]
```diff
@@ -1249,6 +1249,10 @@ def shop_payment_transaction(self, order_id, access_token, **kwargs):
...
+        if order.state == 'cancel':
+            raise ValidationError('The order has been cancelled.')
```
Comment (AntoineVDV) [inline addons/website_sale/controllers/main.py:1254]: Canceled takes only one L in American english.
```suggestion
            raise ValidationError('The order has been canceled.')
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
> The user can still leave to get his wallet when on this page, AFTER it's rendered (and thus after this method is called), and when (s)he comes back the order can be cancelled and the process won't block the payment.

Fair point! It's true that the situation is a bit different here in the eCommerce than when paying through a direct payment link.

Let's then just fix the two issues highlighted below and keep your solution.

File: addons/website_sale/controllers/main.py (hunk 1 above)
Comment (Williambraecky) [reply addons/website_sale/controllers/main.py:1252]: The access check above already checks for the records existence so this isn't really necessary

File: addons/website_sale/controllers/main.py (hunk 2 above)
Comment (ivantodorovich) [reply addons/website_sale/controllers/main.py]: Updated.
It's automatically squashed upon merge, or should I squash commits now?

Comment (AntoineVDV) [reply addons/website_sale/controllers/main.py]: No, you need to squash the commits locally and push the branch.

Comment (ivantodorovich) [reply addons/website_sale/controllers/main.py]: Got it, thanks! It's done

File: addons/website_sale/controllers/main.py [hunk 3]
```diff
@@ -1249,6 +1249,10 @@ def shop_payment_transaction(self, order_id, access_token, **kwargs):
         except AccessError:
             raise ValidationError("The access token is invalid.")
 
+        order = request.env['sale.order'].browse(order_id)
+        if order.state == 'cancel':
+            raise ValidationError('The order has been canceled.')
```
Comment (AntoineVDV) [inline addons/website_sale/controllers/main.py:1254]: `_document_check_access` already returns the **sudoed** record. You can then simply do:
```suggestion
        if order_sudo.state == 'cancel':
            raise ValidationError('The order has been canceled.')
```
Note that the return value of the method is labelled `order_sudo` as a convention to immediately see which recordset's environment is sudoed or not later in the code.

Comment (AntoineVDV) [inline addons/website_sale/controllers/main.py:1254]: The error must be translatable.
```suggestion
            raise ValidationError(_("The order has been canceled."))
```
Note that you'll need to [update the module's translations](https://www.odoo.com/documentation/master/developer/howtos/translations.html#exporting-translatable-term) in this PR.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Sorry for the iterative review; it seems like yesterday was not the best day for me to read code :grimacing:

File: addons/website_sale/controllers/main.py (hunk 3 above)
Comment (ivantodorovich) [reply addons/website_sale/controllers/main.py]: Thanks for the tip ;)

Comment (ivantodorovich) [reply addons/website_sale/controllers/main.py]: Thanks! It's done.
I took the liberty of adding another commit to make the other ValidationError in this method translatable too

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+ rebase-ff

**Pull request closed without merging**